src/macaron/repo_finder/provenance_extractor.py

```diff
@@ -0,0 +1,263 @@
+        elif entry_type.startswith("https://witness.dev/attestations/gitlab/") or entry_type.startswith(
+            "https://witness.dev/attestations/github/"
+        ):
+            repo = json_extract(entry, ["attestation", "projecturl"], str)
+
+    if not commit or not repo:
+        raise ProvenanceExtractionException("Could not extract repo and commit from provenance.")
+
+    return repo, commit
+
+
+def _extract_commit_from_digest_set(digest_set: dict[str, JsonType], valid_algorithms: list[str]) -> str:
+    """Extract the commit from the passed DigestSet.
+
+    The DigestSet is an in-toto object that maps algorithm types to commit hashes (digests).
+    """
+    # TODO decide on a preference for which algorithm to accept.
+    if len(digest_set.keys()) > 1:
+        logger.debug("DigestSet contains multiple algorithms: %s", digest_set.keys())
+
+    for key in digest_set:
+        if key in valid_algorithms:
+            value = digest_set.get(key)
+            if isinstance(value, str):
+                return value
+    raise ProvenanceExtractionException("No valid digest in digest set.")
+
+
+def _clean_spdx(uri: str) -> str:
+    """Clean the passed SPDX URI and return the normalised URL it represents.
+
+    A SPDX URI has the form: git+https://example.com@refs/heads/main
+    """
+    url, _, _ = uri.lstrip("git+").rpartition("@")
+    return url
+
+
+class JsonExtractionException(MacaronError):
+    """When there is an error while extracting from JSON."""
+
+
+T = TypeVar("T", bound=JsonType)
+
+
+def json_extract(entry: dict[str, JsonType], keys: list[str], type_: type[T]) -> T:
+    """Return the value found by following the list of depth-sequential keys inside the passed JSON dictionary.
+
+    The value must be truthy, and be of the passed type.
+
+    Parameters
+    ----------
+    entry: dict[str, JsonType]
+        An entry point into the JSON structure.
+    keys: list[str]
+        The list of depth-sequential keys within the JSON.
+    type: type[T]
+        The type to check the value against and return it as.
+
+    Returns
+    -------
+    T:
+        The found value as the type of the type parameter.
+
+    Raises
+    ------
+    JsonExtractionException
+        Raised if an error occurs while searching for or validating the value.
+    """
+    target = entry
+    for index, key in enumerate(keys):
+        if key not in target:
+            raise JsonExtractionException(f"JSON key not found: {key}")
+        next_target = target[key]
+        if index == len(keys) - 1:
+            if next_target and isinstance(next_target, type_):
+                return next_target
+        else:
+            if not isinstance(next_target, dict):
+                raise JsonExtractionException(f"Cannot extract value from non-dict type: {str(type(next_target))}")
+            target = next_target
+
+    raise JsonExtractionException(f"Failed to find '{' > '.join(keys)}' as type '{type_}' in JSON dictionary.")
```

Comment:
I have a couple of questions as well as suggestions for this function.

### About the final `raise`

I believe the final `raise` only happens in two cases:
* When the extracted value is either falsy or not of the given `type_`, i.e. when the branch `if next_target and isinstance(next_target, type_)` is not taken. This is a type issue.
* When the `keys` list is empty. This seems to me more of a usage issue. Should this function assume that it always takes a non-empty list? If this is the case, then we should clarify it in the docstring at least.

I think the message "Failed to find ... as type ... in JSON dictionary" can be a bit confusing since it does not exactly point out what is the actual problem. The error message could be a bit more specific.

### About the extracted value must be truthy

I wonder why this function requires this. I can see legitimate cases where the value is falsy. An empty list or 0 can be legitimate values.

### About the error message

The error messages in the first two `raise`s give some info about what went wrong, but the info is quite limited. With a hierarchical structure like this, I think it is nice to also include the path to the problematic value in the error messages.

### About the logical flow of this function

I wonder if we could make the logical flow of this function a bit simpler. My intuition initially tells me that the use of the `if index == len(keys) - 1` branching and the `next_target` variable looks a bit clunky.

If I understand it right, there are two phases for this function:

* Traversing down the nested dictionary following the given list of keys.
* Validating the type of the value.

I think we can definitely separate the logic into these two phases which simplifies the implementation.

With all of these points, I would personally implement the function like so:

```suggestion
    target = entry
    target_path = lambda _keys: f".{'.'.join(_keys)}"

    for i, key in enumerate(keys):
        if not isinstance(target, dict):
            raise JsonExtractionException(f"Expect the value {target_path(keys[:i])} to be a dict.")
        if key not in target:
            raise JsonExtractionException(f"JSON key '{key}' not found in {target_path(keys[:i])}.")
        target = target[key]

    if isinstance(target, type_):
        return target
    else:
        raise JsonExtractionException(f"Expect the value {target_path(keys)} to be of type '{type_}'.")
```

I think this is nicer because we got rid of both the `if i == len - 1` and the `next_target` variable. We also introduce a nice invariant for `target` across all iterations that it must initially be a `dict`. I think this is nicer to reason about.